x-pack/platform/plugins/shared/stack_alerts/common/es_query/esql_query_utils.ts

```diff
@@ -110,43 +135,62 @@ export const toEsqlQueryHits = (table: EsqlTable): EsqlQueryHits => {
       },
     },
     rows,
-    cols: getColumnsForPreview(table.columns),
+    cols: isPreview ? getColumnsForPreview(table.columns) : [],
   };
 };
 
-export const toGroupedEsqlQueryHits = (
+export const toGroupedEsqlQueryHits = async (
   table: EsqlTable,
-  alertIdFields: string[]
-): EsqlQueryHits => {
+  alertIdFields: string[],
+  isPreview: boolean = false,
+  chunkSize: number = CHUNK_SIZE
+): Promise<EsqlQueryHits> => {
   const duplicateAlertIds: Set<string> = new Set<string>();
   const longAlertIds: Set<string> = new Set<string>();
   const rows: EsqlDocument[] = [];
   const mappedAlertIds: Record<string, Array<string | null>> = {};
-  const groupedHits = table.values.reduce<Record<string, EsqlHit[]>>((acc, row) => {
-    const document = rowToDocument(table.columns, row);
-    const mappedAlertId = alertIdFields.filter((a) => !isNil(document[a])).map((a) => document[a]);
-    if (mappedAlertId.length > 0) {
-      const alertId = mappedAlertId.join(',');
-      const hit = {
-        _id: ESQL_DOCUMENT_ID,
-        _index: '',
-        _source: document,
-      };
-      if (acc[alertId]) {
-        duplicateAlertIds.add(alertId);
-        acc[alertId].push(hit);
-      } else {
-        acc[alertId] = [hit];
-        mappedAlertIds[alertId] = mappedAlertId;
-      }
-      rows.push({ [ALERT_ID_COLUMN]: alertId, ...document });
+  const groupedHits: Record<string, EsqlHit[]> = {};
+  const numRows = table.values.length;
+  let currentRow = 0;
+
+  const processGroupedEsqlQueryHits = async () => {
+    const chunk = Math.min(currentRow + chunkSize, numRows);
+    for (; currentRow < chunk; currentRow++) {
+      const row = table.values[currentRow];
+      const document = rowToDocument(table.columns, row);
+      const { mappedAlertId, alertId } = getAlertId(document, alertIdFields);
+      if (mappedAlertId.length > 0) {
+        const hit = {
+          _id: ESQL_DOCUMENT_ID,
+          _index: '',
+          _source: document,
+        };
+        if (groupedHits[alertId]) {
+          duplicateAlertIds.add(alertId);
+          groupedHits[alertId].push(hit);
+        } else {
+          groupedHits[alertId] = [hit];
+          mappedAlertIds[alertId] = mappedAlertId;
+        }
+
+        if (isPreview) {
+          rows.push(Object.assign({ [ALERT_ID_COLUMN]: alertId }, document));
 
-      if (mappedAlertId.length >= ALERT_ID_SUGGESTED_MAX) {
-        longAlertIds.add(alertId);
+          if (mappedAlertId.length >= ALERT_ID_SUGGESTED_MAX) {
+            longAlertIds.add(alertId);
+          }
+        }
       }
     }
-    return acc;
-  }, {});
+
+    if (currentRow < numRows) {
+      await new Promise((resolve) => {
```

Comment:
Would it work to do something like

```
for (const row = 0; row < numRows; row++) {
  ...

  if (row % chunkSize === 0) {
    await new Promise(setImmediate);
  }
}
}
```

cc @pmuellr

Reply:
Thanks, this is much cleaner!

Resolved in this commit, https://github.com/elastic/kibana/pull/234689/commits/be5591fef5d0b6407c0f46faa3a11090d26acf3f